XenModel/Actions/Updates/DownloadAndUnzipXenServerPatchAction.cs

```diff
@@ -36,84 +36,228 @@
+                    }
+                }
+
+                log.DebugFormat("Extracting XenServer patch '{0}' and cleaning up completed", updateName);
+            }
+        }
+
+        private void archiveIterator_CurrentFileExtractProgressChanged(long bytesTransferred, long totalBytesToTransfer)
+        {
+            UpdateExtractionProgress(bytesTransferred, totalBytesToTransfer);
+        }
 
-    public class DownloadAndUnzipXenServerPatchAction : AsyncAction, IByteProgressAction
+        protected override void CancelRelatedTask()
+        {
+        }
+
+        protected abstract void UpdateExtractionProgress(long bytesTransferred, long totalBytesToTransfer);
+    }
+
+
+    public class DownloadAndUnzipXenServerPatchAction : DownloadAndUnzipXenServerPatchActionBase, IByteProgressAction
     {
         private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
 
         private const int SLEEP_TIME_TO_CHECK_DOWNLOAD_STATUS_MS = 900;
         private const int SLEEP_TIME_BEFORE_RETRY_MS = 5000;
 
-        //If you consider increasing this for any reason (I think 5 is already more than enough), have a look at the usage of SLEEP_TIME_BEFORE_RETRY_MS in DownloadFile() as well.
+        //If you consider increasing this for any reason (I think 5 is already more than enough),
+        //have a look at the usage of SLEEP_TIME_BEFORE_RETRY_MS in DownloadFile() as well.
         private const int MAX_NUMBER_OF_TRIES = 5;
 
-        private readonly Uri address;
-        private readonly string outputFileName;
-        private readonly string updateName;
-        private readonly string[] updateFileSuffixes;
-        private readonly bool downloadUpdate;
         private readonly bool skipUnzipping;
+        private readonly Uri _patchUri;
         private DownloadState patchDownloadState;
         private Exception patchDownloadError;
+        private WebClient client;
 
-        public string PatchPath { get; private set; }
 
-        public string ByteProgressDescription { get; set; }
+        public DownloadAndUnzipXenServerPatchAction(string patchName, Uri uri, params string[] updateFileExtensions)
+            : base(patchName, updateFileExtensions)
+        {
+            Title = string.Format(Messages.DOWNLOAD_AND_EXTRACT_ACTION_TITLE, patchName);
+            _patchUri = uri;
+            skipUnzipping = updateFileSuffixes.Any(item => _patchUri.ToString().EndsWith(item));
+        }
 
-        public DownloadAndUnzipXenServerPatchAction(string patchName, Uri uri, string outputFileName, bool suppressHist,
-            params string[] updateFileExtensions)
-            : base(null, uri == null
-                ? string.Format(Messages.UPDATES_WIZARD_EXTRACT_ACTION_TITLE, patchName)
-                : string.Format(Messages.DOWNLOAD_AND_EXTRACT_ACTION_TITLE, patchName), string.Empty, suppressHist) 
+        public override void RecomputeCanCancel()

```

Comment:
Dont see what the point in this function is, it assigns a variable that is never used

Reply:
This is base class infrastructure used to decide whether an action is cancellable (e.g. if the action were shown on a dialog, it would decide whether to enable the cancel button for the user).